google/cloud/status_or.h

```diff
@@ -24,33 +24,54 @@ namespace google {
 namespace cloud {
 inline namespace GOOGLE_CLOUD_CPP_NS {
 /**
- * `StatusOr<T>` is used to return a value or an error status.
+ * Holds a value or a `Status` indicating why there is no value.
  *
- * If the library is compiled with exceptions disabled there is no standard C++
- * mechanism to report errors. In that case we use this class to wrap the values
- * returned to the application.
- *
- * The application typically calls an API in the library and must check the
- * returned error status:
+ * `StatusOr<T>` represents either a usable `T` value or a `Status` object
+ * explaining why a `T` value is not present. Typical usage of `StatusOr<T>`
+ * looks like usage of a smart pointer, or even a std::optional<T>, in that you
+ * first check its validity using a conversion to bool (or by calling
+ * `StatusOr<T>::ok()`), then you may dereference the object to access the
+ * contained value. It is undefined behavior (UB) to dereference a
+ * `StatusOr<T>` that is not "ok". For example:
  *
  * @code
- * namespace gcs = google::cloud::storage;
- * void AppCode(gcs::noex::Client client) {
- *   gcs::StatusOr<gcs::BucketMetadata> meta_err = client.GetBucketMetadata(
- *       "my-bucket-name");
- *   if (!meta_err) {
- *       std::cerr << "Error in GetBucketMetadata: " << meta_err.status()
- *                 << std::endl;
- *       return;
- *   }
- *   gcs::BucketMetadata meta = *meta_err;
- *   // Do useful work here.
+ * StatusOr<Foo> foo = FetchFoo();
+ * if (!foo) {  // Same as !foo.ok()
```

Comment:
Could you flip the comment and code used in the segment? 

-  s/`if (!foo) {`/`if (!foo.ok()) {`/
-  s/`Same as !foo.ok()`/`Same as !foo`/ 

I'd like to keep the examples and documented snippet consistent with how they check status.

Reply:
FTR, I write it this way to encourage the `!foo` syntax. Over time within Google we've moved to preferring this form over `!foo.ok`, because the former syntax works exactly the same with:

- std::optional
- std::unique_ptr
- and any T*

so it supports the mental model of thinking about a StatusOr as a smart pointer, which I added to the documentation above.

I'll still change, but I wanted to explain why I specifically changed to and prefer the existing form.

Reply:
Thanks @devjgm if that's the case. Do you think the examples should use this model instead?

Reply:
Yes, I do. Since the header contains the canonical docs, I thought it made sense to update these first. But I do think that examples should generally use the `if (!foo)` form. And we should not use `.value()` if we already checked the object's validity. In that case, we should instead use the dot or arrow operator (as is shown in this example). We should probably only use .value() if we're relying on it throwing for us, which will never happen if we explicitly checked the validity.

Reply:
thanks.